### PR TITLE
Support ExtraPrefixes in ConnectionContext #572

### PR DESCRIPTION
**Pull request by haiodo**
Support ExtraPrefixes in ConnectionContext

Now it is possible to pass a request for allocation of subnet prefixes and easy NSE prefix management.

For NSE:
```go
pool, _ := prefix_pool.NewPrefixPool("10.20.1.0/24")
myExtraPrefixRequest := []connectioncontext.ExtraPrefixRequest { &connectioncontext.ExtraPrefixRequest{
		RequiredNumber:  5,
		RequestedNumber: 10,
		PrefixLen:       32,
		AfFamily:        &connectioncontext.IpFamily{Family: connectioncontext.IpFamily_IPV4,},
	}
}
srcIP, dstIP, prefixes, err := pool.Extract("c1", connectioncontext.IpFamily_IPV4, myExtraPrefixRequest...)
if err != nil {
  return nil, err
}
// So we got all srcIP/dstIP and prefixes so we could easy return Connection with all stuff set.

```
Also pool remember connection to prefixes mapping so 
```go
pool.Release("c1")
``` 
will return prefixes back.

And of cause generic:

```go

// Extract prefixes
func ExtractPrefixes(prefixes []string, requests ...*connectioncontext.ExtraPrefixRequest) (requested []string, remaining []string, err error) {}

// Release prefixes
func ReleasePrefixes(prefixes []string, released ...string) (remaining []string, err error) {} 

// Get next IP
func IncrementIP(sourceIp net.IP, ipNet *net.IPNet) (net.IP, error) {} 

// Range of addresses
func AddressRange(network *net.IPNet) (net.IP, net.IP) {} 

// Number of addresses
func AddressCount(prefixes ...string) uint64 {} 
```

could be used

Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

### CODE REVIEW

**Comment by haiodo** [comment]
https://github.com/haiodo/networkservicemesh/pull/new/icmpResponderNSEprefixPool is example of ICMP responder NSE, will create PR after this one will be merged.

File: controlplane/pkg/apis/connectioncontext/connectioncontext.proto [hunk 1]
```diff
@@ -30,6 +30,23 @@ message Route {
...
+        IPV4 = 0;
+        IPV6 = 1;
+//        ALL = 2;
+//        MPLS = 3;
+    }
+    Family family = 1;
+}
+
+message ExtraPrefixRequest {
+    IpFamily Af_family      = 1;
```
Comment (nickolaev) [inline controlplane/pkg/apis/connectioncontext/connectioncontext.proto]: Start with low case `af_family`

Comment (nickolaev) [reply controlplane/pkg/apis/connectioncontext/connectioncontext.proto]: or even better `addr_family` as ultimately AF stands for Address Family.

File: controlplane/pkg/apis/connectioncontext/connectioncontext.proto [hunk 2]
```diff
@@ -30,6 +30,23 @@ message Route {
     string prefix = 1;                  /* destination address + prefix in format <address>/<prefix> */
 }
 
+message IpFamily {
+    enum Family {
+        IPV4 = 0;
+        IPV6 = 1;
+//        ALL = 2;
```
Comment (nickolaev) [inline controlplane/pkg/apis/connectioncontext/connectioncontext.proto]: Not sure what ALL and MPLS has to do here, please remove.

File: controlplane/pkg/apis/connectioncontext/connectioncontext.proto [hunk 3]
```diff
@@ -40,5 +57,8 @@ message ConnectionContext {
     repeated Route routes = 5;          /* a list of routes via dst_ip_addr as gateway */
     repeated string excluded_prefixes = 6; /* a list of excluded passed to NSE */
 
-    repeated IpNeighbor ip_neighbors  = 7;
+    repeated IpNeighbor ip_neighbors  = 7; /* A list of ARP table neighbours addresses */
+
+    repeated ExtraPrefixRequest extra_prefix_request = 8;   /* A request for NSE to have provide extra prefixes */
```
Comment (nickolaev) [inline controlplane/pkg/apis/connectioncontext/connectioncontext.proto]: An extra "have" here. Plese replace with "A request for NSE to provide extra prefixes".

File: controlplane/pkg/apis/connectioncontext/helpers.go [hunk 4]
```diff
@@ -45,3 +45,42 @@ func (c *ConnectionContext) MeetsRequirements(original *ConnectionContext) error
...
 	return nil
 }
+
+func (c *ExtraPrefixRequest) IsValid() error {
+	if c == nil {
+		return fmt.Errorf("ExtraPrefixRequest should not be nil...")
+	}
+
+	if c.RequiredNumber < 1 {
+		return fmt.Errorf("ExtraPrefixRequest.RequiredNumber should be positive number >=1: %v", c)
```
Comment (nickolaev) [inline controlplane/pkg/apis/connectioncontext/helpers.go]: No need to put the full `ExtraPrefixRequest.RequiredNumber` in this and consequent messages. Please replace with simple `RequiredNumber`. Please clean all the messages below too.

File: controlplane/pkg/prefix_pool/prefixpool.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+	// Check if requests are valid.
+
+	for _, request := range requests {
+		error := request.IsValid()
+		if error != nil {
+			return nil, prefixes, error
+		}
+	}
+
+	//1 find prefix of disired prefix len and return it
```
Comment (nickolaev) [inline controlplane/pkg/prefix_pool/prefixpool.go]: 'desired'

File: controlplane/pkg/apis/connectioncontext/helpers.go (hunk 4 above)
Comment (haiodo) [reply controlplane/pkg/apis/connectioncontext/helpers.go]: This is done to be consistent with other helpers.go code.

https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/local/connection/helpers.go and others.

**Pull request merged**